storagenode/storagenodedb/database.go

```diff
@@ -115,108 +116,80 @@ func New(log *zap.Logger, config Config) (*DB, error) {
+	// that each uses internally to do data access to the SQLite3 databases.
+	// The reason it was done this way was because there's some outside consumers that are
+	// taking a reference to the business object.
+	legacyInfoDB, err := db.openDatabase(LegacyInfoDBName)
 	if err != nil {
-		return nil, err
-	}
-
-	db := &DB{
-		log:    log,
-		pieces: pieces,
-		kdb:    teststore.New(),
-		ndb:    teststore.New(),
-		adb:    teststore.New(),
-
-		// Initialize databases. Currently shares one info.db database file but
-		// in the future these will initialize their own database connections.
-		versionsDB:        newVersionsDB(versionsDB, versionsPath),
-		v0PieceInfoDB:     newV0PieceInfoDB(versionsDB, versionsPath),
-		bandwidthDB:       newBandwidthDB(versionsDB, versionsPath),
-		ordersDB:          newOrdersDB(versionsDB, versionsPath),
-		pieceExpirationDB: newPieceExpirationDB(versionsDB, versionsPath),
-		pieceSpaceUsedDB:  newPieceSpaceUsedDB(versionsDB, versionsPath),
-		reputationDB:      newReputationDB(versionsDB, versionsPath),
-		storageUsageDB:    newStorageusageDB(versionsDB, versionsPath),
-		usedSerialsDB:     newUsedSerialsDB(versionsDB, versionsPath),
-		satellitesDB:      newSatellitesDB(versionsDB, versionsPath),
+		return errs.Combine(err, db.closeDatabases())
 	}
-	return db, nil
+	db.legacyInfoDB.Configure(legacyInfoDB)
+	db.bandwidthDB.Configure(legacyInfoDB)
+	db.ordersDB.Configure(legacyInfoDB)
+	db.pieceExpirationDB.Configure(legacyInfoDB)
+	db.v0PieceInfoDB.Configure(legacyInfoDB)
+	db.pieceSpaceUsedDB.Configure(legacyInfoDB)
+	db.reputationDB.Configure(legacyInfoDB)
+	db.storageUsageDB.Configure(legacyInfoDB)
+	db.usedSerialsDB.Configure(legacyInfoDB)
+	db.satellitesDB.Configure(legacyInfoDB)
+	return nil
 }
 
 // openDatabase opens or creates a database at the specified path.
-func openDatabase(path string) (*sql.DB, error) {
+func (db *DB) openDatabase(dbName string) (*sql.DB, error) {
+	path := filepath.Join(db.dbDirectory, db.FilenameFromDBName(dbName))
 	if err := os.MkdirAll(filepath.Dir(path), 0700); err != nil {
 		return nil, err
 	}
 
-	db, err := sql.Open("sqlite3", "file:"+path+"?_journal=WAL&_busy_timeout=10000")
-	if err != nil {
-		return nil, ErrDatabase.Wrap(err)
-	}
-
-	dbutil.Configure(db, mon)
-	return db, nil
-}
-
-// // openTestDatabase creates an in memory database.
-func openTestDatabase() (*sql.DB, error) {
-	// create memory DB with a shared cache and a unique name to avoid collisions
-	db, err := sql.Open("sqlite3", fmt.Sprintf("file:memdb%d?mode=memory&cache=shared", rand.Int63()))
+	sqlDB, err := sql.Open("sqlite3", "file:"+path+"?_journal=WAL&_busy_timeout=10000")
 	if err != nil {
 		return nil, ErrDatabase.Wrap(err)
 	}
 
-	// Set max idle and max open to 1 to control concurrent access to the memory DB
-	// Setting max open higher than 1 results in table locked errors
-	db.SetMaxIdleConns(1)
-	db.SetMaxOpenConns(1)
-	db.SetConnMaxLifetime(-1)
+	db.sqlDatabases[dbName] = sqlDB
```

Comment:
Because this has no mutex (which is fine), I would prefer a comment above the method that states it is not concurrent-safe. And that propagates all the way up to the `New` call.

It's a small thing, but I think it's nice to be explicit about these guarantees (and the lack of them) when possible.

Reply:
Done!